### ISSUE TITLE
error while loading shared libraries: libxml2.so.2 (liblouisutdml)

### ISSUE THREAD

**Issue by josteinaj** [nlbdev/pipeline#223]
```
2019-05-22 11:11:17,449 [DEBUG] com.xmlcalabash.library.DefaultStep - Running {http://liblouis.org/liblouis}translate-file !1.11
2019-05-22 11:11:17,450 [DEBUG] o.d.p.b.l.i.LiblouisutdmlProcessBuilderImpl - 
/opt/daisy-pipeline2/data/felix-cache/bundle87/data/resources0/x86_64/liblouisutdml/file2brl
	-f
	/opt/daisy-pipeline2/data/felix-cache/bundle86/data/resources0/nemeth.cfg
	-ClineEnd=\n
	-CpageEnd=\f
	/opt/daisy-pipeline2/data/jobs/bbaedd39-ab0d-458e-a1c4-8bfc06aae98a/output/temp-dir/temp9010815424262790836/convert/liblouisutdml.16218723888448991415.xml
	/opt/daisy-pipeline2/data/jobs/bbaedd39-ab0d-458e-a1c4-8bfc06aae98a/output/temp-dir/temp9010815424262790836/convert/liblouisutdml.17035988027489962187.txt
2019-05-22 11:11:17,457 [DEBUG] o.d.p.b.l.i.LiblouisutdmlProcessBuilderImpl - 
exit value: 127
2019-05-22 11:11:17,461 [ERROR] o.d.p.b.l.i.LiblouisutdmlProcessBuilderImpl - Error during liblouisutdml conversion
java.lang.RuntimeException: liblouisutdml exited with value 127
	at org.daisy.pipeline.braille.liblouis.impl.LiblouisutdmlProcessBuilderImpl.translateFile(LiblouisutdmlProcessBuilderImpl.java:161) ~[na:na]
	at org.daisy.pipeline.braille.liblouis.calabash.impl.TranslateFileStep.run(TranslateFileStep.java:242) ~[na:na]
	at com.xmlcalabash.runtime.XAtomicStep.run(XAtomicStep.java:396) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at com.xmlcalabash.extensions.Eval.run(Eval.java:248) ~[na:na]
	at org.daisy.pipeline.braille.common.calabash.impl.PxTransformStep.run(PxTransformStep.java:139) ~[na:na]
	at com.xmlcalabash.runtime.XAtomicStep.run(XAtomicStep.java:396) ~[na:na]
	at com.xmlcalabash.runtime.XViewport.processStartElement(XViewport.java:156) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:199) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:188) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.match(ProcessMatch.java:95) ~[na:na]
	at com.xmlcalabash.runtime.XViewport.run(XViewport.java:99) ~[na:na]
	at com.xmlcalabash.runtime.XCompoundStep.run(XCompoundStep.java:264) ~[na:na]
	at com.xmlcalabash.runtime.XGroup.run(XGroup.java:30) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at com.xmlcalabash.runtime.XPipelineCall.run(XPipelineCall.java:101) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at org.daisy.common.xproc.calabash.impl.CalabashXProcPipeline.run(CalabashXProcPipeline.java:251) ~[na:na]
	at org.daisy.pipeline.job.Job.run(Job.java:218) ~[na:na]
	at org.daisy.pipeline.job.impl.DefaultJobExecutionService$1.run(DefaultJobExecutionService.java:111) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
2019-05-22 11:11:17,461 [DEBUG] o.d.p.b.l.i.LiblouisutdmlProcessBuilderImpl - 
stderr:
	/opt/daisy-pipeline2/data/felix-cache/bundle87/data/resources0/x86_64/liblouisutdml/file2brl: error while loading shared libraries: libxml2.so.2: cannot open shared object file: No such file or directory
2019-05-22 11:11:17,461 [DEBUG] o.a.h.i.c.PoolingHttpClientConnectionManager - Connection manager is shutting down
2019-05-22 11:11:17,465 [DEBUG] o.a.h.i.c.PoolingHttpClientConnectionManager - Connection manager shut down
2019-05-22 11:11:17,466 [INFO ] org.daisy.pipeline.job.Job - Changing job status to: ERROR
2019-05-22 11:11:17,466 [INFO ] o.d.p.p.impl.job.PersistentJob - Changing Status:ERROR
2019-05-22 11:11:17,466 [DEBUG] o.d.p.p.impl.job.PersistentJob - Updating object
2019-05-22 11:11:17,467 [DEBUG] o.d.p.p.i.job.PersistentJobContext - load script XProcScript[name=EPUB 3 til PEF (NLB)]
2019-05-22 11:11:17,467 [DEBUG] o.d.p.p.i.job.PersistentJobContext - load script XProcScript[name=EPUB 3 til PEF (NLB)]
2019-05-22 11:11:17,468 [DEBUG] o.d.p.p.i.job.PersistentJobContext - Post loading jobcontext
2019-05-22 11:11:17,468 [DEBUG] o.d.p.p.i.job.PersistentJobContext - load script XProcScript[name=EPUB 3 til PEF (NLB)]
2019-05-22 11:11:17,471 [DEBUG] o.e.persistence.logging.default - Connection(157722686)--Thread(Thread[Thread-24,5,Configuration Admin Service])--UPDATE jobs SET STATUS = ? WHERE (job_id = ?)
	bind => [2 parameters bound]
2019-05-22 11:11:17,485 [DEBUG] org.daisy.pipeline.push.impl.Poster - Status changed bbaedd39-ab0d-458e-a1c4-8bfc06aae98a->ERROR
2019-05-22 11:11:17,487 [DEBUG] o.e.persistence.logging.default - Connection(905122962)--Thread(Thread[Thread-24,5,Configuration Admin Service])--SELECT job_id, PRIORITY, STATUS, CONTEXT_job_id FROM jobs WHERE (job_id = ?)
	bind => [1 parameter bound]
2019-05-22 11:11:17,488 [DEBUG] o.d.p.p.i.job.PersistentJobContext - load script XProcScript[name=EPUB 3 til PEF (NLB)]
2019-05-22 11:11:17,489 [DEBUG] o.d.p.p.i.job.PersistentJobContext - load script XProcScript[name=EPUB 3 til PEF (NLB)]
2019-05-22 11:11:17,490 [DEBUG] o.d.p.p.i.job.PersistentJobContext - Post loading jobcontext
2019-05-22 11:11:17,490 [ERROR] org.daisy.pipeline.job.Job - job finished with error state
java.lang.RuntimeException: Error during liblouisutdml conversion
	at translate-mathml.xpl:30
	at {http://liblouis.org/liblouis}translate-file(translate-mathml.xpl:30)
	at {http://www.daisy.org/ns/pipeline/xproc}transform(epub3-to-pef.convert.xpl:190)
	at epub3-to-pef.convert.xpl:189
	at epub3-to-pef.convert.xpl:184
	at {http://www.daisy.org/ns/pipeline/xproc}epub3-to-pef(__processed__epub3-to-pef.xpl:429)
2019-05-22 11:11:17,492 [DEBUG] org.daisy.pipeline.job.Job - job finished with error state
org.daisy.common.xproc.XProcErrorException: java.lang.RuntimeException: Error during liblouisutdml conversion
	at org.daisy.common.xproc.calabash.impl.CalabashXProcPipeline.run(CalabashXProcPipeline.java:255) ~[na:na]
	at org.daisy.pipeline.job.Job.run(Job.java:218) ~[na:na]
	at org.daisy.pipeline.job.impl.DefaultJobExecutionService$1.run(DefaultJobExecutionService.java:111) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: com.xmlcalabash.core.XProcException$3: java.lang.RuntimeException: Error during liblouisutdml conversion
	at com.xmlcalabash.core.XProcException.rebaseOnto(XProcException.java:336) ~[na:na]
	at com.xmlcalabash.core.XProcException.rebaseOnto(XProcException.java:305) ~[na:na]
	at com.xmlcalabash.runtime.XStep.handleException(XStep.java:313) ~[na:na]
	at com.xmlcalabash.runtime.XPipelineCall.run(XPipelineCall.java:103) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at org.daisy.common.xproc.calabash.impl.CalabashXProcPipeline.run(CalabashXProcPipeline.java:251) ~[na:na]
	... 3 common frames omitted
Caused by: java.lang.RuntimeException: Error during liblouisutdml conversion
	at org.daisy.pipeline.braille.liblouis.impl.LiblouisutdmlProcessBuilderImpl.translateFile(LiblouisutdmlProcessBuilderImpl.java:165) ~[na:na]
	at org.daisy.pipeline.braille.liblouis.calabash.impl.TranslateFileStep.run(TranslateFileStep.java:242) ~[na:na]
	at com.xmlcalabash.runtime.XAtomicStep.run(XAtomicStep.java:396) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at com.xmlcalabash.extensions.Eval.run(Eval.java:248) ~[na:na]
	at org.daisy.pipeline.braille.common.calabash.impl.PxTransformStep.run(PxTransformStep.java:139) ~[na:na]
	at com.xmlcalabash.runtime.XAtomicStep.run(XAtomicStep.java:396) ~[na:na]
	at com.xmlcalabash.runtime.XViewport.processStartElement(XViewport.java:156) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:199) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:233) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.traverse(ProcessMatch.java:188) ~[na:na]
	at com.xmlcalabash.util.ProcessMatch.match(ProcessMatch.java:95) ~[na:na]
	at com.xmlcalabash.runtime.XViewport.run(XViewport.java:99) ~[na:na]
	at com.xmlcalabash.runtime.XCompoundStep.run(XCompoundStep.java:264) ~[na:na]
	at com.xmlcalabash.runtime.XGroup.run(XGroup.java:30) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.doRun(XPipeline.java:236) ~[na:na]
	at com.xmlcalabash.runtime.XPipeline.run(XPipeline.java:136) ~[na:na]
	at com.xmlcalabash.runtime.XPipelineCall.run(XPipelineCall.java:101) ~[na:na]
	... 6 common frames omitted
Caused by: java.lang.RuntimeException: liblouisutdml exited with value 127
	at org.daisy.pipeline.braille.liblouis.impl.LiblouisutdmlProcessBuilderImpl.translateFile(LiblouisutdmlProcessBuilderImpl.java:161) ~[na:na]
	... 26 common frames omitted
```

libxml2 should be installed in the docker image already:

```Dockerfile
RUN apt-get update && apt-get install -y libxml2 libxml2-dev libxml2-utils libxslt1-dev zlib1g-dev liblzma-dev
```

The engine is currently not running outside of docker for me (there's a lot of "Service factory returned null" exceptions), so I don't know if it's got to do with how libxml2 is installed in the docker image or if it's got to do with how liblouisutdml tries to locate libxml2.

**Comment by josteinaj** [comment]
By the way, to reproduce: use the test EPUB from the nordic migrator and try producing a PEF with the NLB-script.

The docker image `nlbdev/pipeline:2019-05-22` can be used for testing.

**Comment by josteinaj** [comment]
Oh, libxml2 is installed in the builder, but not in the actual image. That's probably the issue. Will try.

**Comment by bertfrees** [comment]
Or is it possible that it is looking for a "libxml2.so.2" but that the name has changed (to libxml2.so.3 for instance)?

Regarding the "Service factory returned null" exceptions, are you sure this is not a Java version issue? Several people have reported this issue, but in all of the cases using the AdoptOpenJDK version fixed it.

**Comment by josteinaj** [comment]
`find /usr/lib | grep libxml` returns nothing. It's probably just not installed.

Service factory: I changed JAVA_HOME to AdoptOpenJDK 11, but I'll have a closer look.

**Comment by josteinaj** [comment]
Fixed in 5accd0bbda15023b63ff87cd4ceed9c9ecdea46b.

**service factory error:**

```bash
➜  ~ java -version
openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.3+7, mixed mode)
➜  ~ echo $JAVA_HOME
/opt/jdk-11.0.3+7
```

Log with errors attached: 
[service-factory-returned-null.log](https://github.com/nlbdev/pipeline/files/3207507/service-factory-returned-null.log)
